roles/kubernetes/preinstall/handlers/main.yml

```diff
@@ -32,4 +36,35 @@
   service:
     name: kubelet
     state: restarted
+  notify:
+    - Preinstall | kube-controller configured
+    - Preinstall | remove controller
+    - Preinstall | pause for controller
+    - Preinstall | restart controller
   when: "{{ inventory_hostname in groups['kube-master'] and not dns_early|bool }}"
+
+#
+# The following is used to restart the kube-controller to pick up
+# DNS changes in host_resolvconf mode.  This should be done much "better"
+# This is what seems to work now though.  This is only used when
+# kube-controller manifest is present and we are using host_resolvconf.
+#
+# This is triggered by a kubelet restart or network restart.
+#
+- name: Preinstall | kube-controller configured
+  stat: path="{{ kube_manifest_dir }}/kube-controller-manager.manifest"
+  register: kube_controller_set
+  when: "{{ inventory_hostname in groups['kube-master'] and dns_mode != 'none' and resolvconf_mode == 'host_resolvconf' }}"
+
+- name: Preinstall | remove controller
+  command: removes="{{ kube_manifest_dir }}/kube-controller-manager.manifest" creates=/tmp/jj mv "{{ kube_manifest_dir }}/kube-controller-manager.manifest" /tmp/jj
```

Comment:
In my experiments with forcing restart of kube-apiserver, this wasn't the most reliable method. The most reliable method of killing a static pod is to change its image to an invalid value. Look here: https://github.com/kubernetes-incubator/kargo/blob/master/roles/kubernetes/master/tasks/pre-upgrade.yml#L37